pkg/state/steps_uki.go

```diff
@@ -611,3 +615,100 @@ func (s *State) ExtractCerts(g *herd.Graph, opts ...herd.OpOption) error {
 		return nil
 	}))...)
 }
+
+// TransitionSysext is a workaround for upgrades from 3.3.x to 3.4.x
+// In 3.3.x we had the extensions in the EFI dir directly, under /efi/EFI/kairos/{active,passive}.efi.extra.d/
+// In 3.4.x we moved them to /var/lib/kairos/extensions/ for generic and for enabled ones to /var/lib/kairos/extensions/{active,passive}/
+// This is a workaround to move the extensions from the old location to the new one to help with upgrades
+// The order is:
+// Check both active and passive dirs
+// If something is found, move it to the new location at /var/lib/kairos/extensions/
+// Enable it by creating a softlink from /var/lib/kairos/extensions/{active,passive}/EXTENSION to /var/lib/kairos/extensions/EXTENSION
+// Remove it from the old location
+func (s *State) TransitionSysext(g *herd.Graph, opts ...herd.OpOption) error {
+	return g.Add(cnst.OpUkiTransitionSysext, append(opts, herd.WithCallback(func(_ context.Context) error {
+		if !state.EfiBootFromInstall(internalUtils.Log) {
+			internalUtils.Log.Debug().Msg("Not transitioning sysext as we think we are booting from removable media")
+			return nil
+		}
+
+		// Check or create target dir
+		if _, err := os.Stat(s.path("/var/lib/kairos/extensions")); os.IsNotExist(err) {
+			err = os.MkdirAll(s.path("/var/lib/kairos/extensions"), 0755)
+			if err != nil {
+				return err
+			}
+		}
+
+		// We have to remount the EFI partition as RW to be able to move the files
+		err := syscall.Mount(cnst.EfiDir, cnst.EfiDir, cnst.UkiDefaultEfiimgFsType, syscall.MS_REMOUNT, "rw")
+		if err != nil {
+			internalUtils.Log.Err(err).Msg("Mounting EFI partition")
+			return err
+		}
+		// We need to remount it as RO after we are done
+		defer func(source string, target string, fstype string, flags uintptr, data string) {
+			err := syscall.Mount(source, target, fstype, flags, data)
+			if err != nil {
+				internalUtils.Log.Err(err).Msg("Mounting EFI partition as RO")
+			} else {
+				internalUtils.Log.Debug().Msg("Remounting EFI partition as RO")
+			}
+		}(cnst.EfiDir, cnst.EfiDir, cnst.UkiDefaultEfiimgFsType, syscall.MS_REMOUNT|syscall.MS_RDONLY, "")
```

Comment:
since this is not a re-usable function, I don't see a reason why you can't just hardcode the `syscall.Mount` flags in the function instead of passing them through the defer call.

Reply:
oh I get it.